### ISSUE TITLE
Summary: Improving MapReduce Performance in Heterogeneous Environments

### ISSUE THREAD

**Issue by ahsmha** [tapaswenipathak/linux-kernel-stats#85]

**Comment by ahsmha** [comment]
https://www.usenix.org/legacy/events/osdi08/tech/full_papers/zaharia/zaharia.pdf

**Comment by ahsmha** [comment]
This paper discusses the problem of node heterogeneity in distributed computing systems and how it affects the performance of the MapReduce programming model. The authors analyze the problem of "speculative execution" in MapReduce, which is a technique used to speed up the completion of a computation by running multiple copies of a task on different nodes, and then using the results of the fastest one. However, the existing threshold-based scheduling algorithm in Hadoop and progress-rate-based algorithms in general have some flaws, they are not robust to variations in node performance and do not effectively utilize idle resources.

To address these issues, the authors designed a new scheduling algorithm called LATE (Longest Approximate Time to End) which uses estimated finish times to speculatively execute the tasks that hurt the response time the most. The LATE algorithm estimates the finish times of tasks based on their progress so far and the performance of the node on which they are running. It then uses this information to identify the tasks that are likely to take the longest to complete, and runs additional copies of these tasks on other nodes. The goal of LATE is to speed up the overall computation by completing the slowest tasks as quickly as possible.

The authors discuss a method for estimating the time left for a task based on the progress score provided by Hadoop, as (1 − ProgressScore)/ProgressRate. However, they acknowledge that this method can sometimes backfire and incorrectly estimate that a task that was launched later than an identical task will finish earlier. This happens when a task has two phases in which it runs at different rates, and the task slows down in its second phase, making "younger" tasks seem faster. The authors note that this situation does not frequently arise in typical MapReduce jobs in Hadoop, but that it is an interesting and subtle problem in scheduling using progress rates.

The authors evaluated LATE on real workloads on Amazon's Elastic Compute Cloud and found that it performed significantly better than Hadoop's default speculative execution algorithm. They also compare LATE with other related work and found that it is a simple, robust scheduling algorithm that addresses the problem of node heterogeneity in distributed computing systems.

In summary, this paper describes a new scheduling algorithm called LATE that addresses the problem of node heterogeneity in distributed computing systems. LATE uses estimated finish times to speculatively execute the tasks that hurt the response time the most and is able to significantly speed up the overall computation. The authors have evaluated this algorithm on real workloads and found that it outperforms existing algorithms.

**Comment by KavitaMeena23** [comment]
**mention the following in your summary**
1. brief scheduling in Hadoop (section 2)
2. include advantages of LATE Scheduler (section 4.1)
3. brief the experiments done to evaluate the performance of the scheduler (section 5)
4. related work (section 7)

**Comment by KavitaMeena23** [comment]
**Reviewed**

**Comment by duttabhishek0** [comment]
@tapaswenipathak
